lib/vm/src/instance.rs

```diff
@@ -978,6 +1052,97 @@ impl InstanceHandle {
         Ok(handle)
     }
 
+    /// Create a new `InstanceHandle` pointing at the instance
+    /// pointed to by the given `VMContext` pointer.
+    ///
+    /// # Safety
+    /// This is unsafe because it doesn't work on just any `VMContext`, it must
+    /// be a `VMContext` allocated as part of an `Instance`.
+    pub(crate) unsafe fn from_vmctx(vmctx: *const VMContext) -> Self {
+        /*
+        let instance = (&*vmctx).instance();
+
+        Self {
+            instance: Arc::from_raw(instance as *const InstanceAllocator),
+            // We consider `vmctx` owns the `Instance`, not `Self`.
+            //owned_instance: false,
+        }
+        */
+        todo!()
+    }
+
+    /// Return a reference to the contained `Instance`.
+    pub(crate) fn instance(&self) -> &InstanceAllocator {
+        &self.instance
+    }
+
+    /// Get the locations of where the local `VMMemoryDefinition`s should be stored.
+    ///
+    /// This function lets us create `Memory` objects on the host with backing
+    /// memory in the VM.
+    ///
+    /// # Safety
+    /// - `instance_ptr` must point to enough memory that all of the offsets in
+    ///   `offsets` point to valid locations in memory.
+    pub unsafe fn memory_definition_locations(
+        instance_ptr: NonNull<u8>,
+        offsets: &VMOffsets,
+    ) -> Vec<NonNull<VMMemoryDefinition>> {
+        // `NonNull<u8>` here actually means `NonNull<Instance>`. See
+        // `Self::allocate_instance` to understand why.
+        let instance_ptr: NonNull<Instance> = instance_ptr.cast();
```

Comment:
Doesn't this cast make the `add` below incorrect?  The logic is written in terms of bytes and ptr::add multiplies by the size.

Thanks for calling out that you made changes here or I never would have caught this.  In the future it's probably best to avoid changing moved code in a PR this big and complex: github's diff viewer makes understanding these types of changes really difficult.

Reply:
Indeed, https://doc.rust-lang.org/std/primitive.pointer.html#method.add:

> `count` is in units of `T`; e.g., a `count` of 3 represents a pointer offset of `3 * size_of::<T>()` bytes.

That's probably why memories and tables are failing. I'm fixing that! Thanks for noticing!